### DISCUSSION TITLE
How to check status of previous step?

### DISCUSSION THREAD

**Discussion by martyrs** [community/community#26060] [Actions]
<p>Hello,<br>
I have a workflow that builds docker image and next action is creating comment on each workflow run.<br>
However, I need to check if build step has failed or nott and take action.</p>
<p>Please note, in my build.sh script, i already set <code>set -e</code> to prevent the execution of bash script if i get any errors.</p>
<p>The following workflow doesn’t work for some reason, I think i am unable to read <code>exit_code</code>. Any help would be much appreciated.<br>
Thanks.</p>
<pre><code class="lang-auto">
      - name: Build Docker Image
        id: build
        continue-on-error: true
        run: |
          ./build.sh
          echo ::set-output name=exit_code::$?
          echo "Exited with $?"
        env:
          BUILD_ACTOR: ${{ github.event.client_payload.actor }}
          BUILD_INITIATOR: 'cmd'
          GITHUB_REF: $GITHUB_REF

      - name: check steps.build.outputs.exit_code
        run: echo ${{ steps.build.outputs.exit_code }}
        if: always()

      - name: Create comment if build is ok
        if: steps.build.outputs.exit_code == 0
        uses: peter-evans/create-or-update-comment@v1
        with:
          token: ${{ secrets.ACCESS_TOKEN }}
          repository: ${{ github.event.client_payload.repository }}
          issue-number: ${{ github.event.client_payload.issue_id }}
          body : |
            Docker image is now ready at

            ${{ github.event.client_payload.image_version }}

      - name: Create comment if build fails
        if: steps.build.outputs.exit_code != 0
        uses: peter-evans/create-or-update-comment@v1
        with:
          token: ${{ secrets.ACCESS_TOKEN }}
          repository: ${{ github.event.client_payload.repository }}
          issue-number: ${{ github.event.client_payload.issue_id }}
          body : |
            Docker build has been failed.

            [Debug output of workflow][1]

            [1]: ${{ steps.vars.outputs.run-url }}
</code></pre>

**Comment by airtower-luna** [comment, marked as answer]
<p>It’s hard to say exactly what is causing your problem without a detailed description of the problem (and ideally a link to a workflow run), but I already see a few problems:</p>
<aside class="quote no-group" data-username="martyrs" data-post="1" data-topic="128757">
<div class="title">
<div class="quote-controls"></div>
 martyrs:</div>
<blockquote>
<pre><code class="lang-auto">        run: |
          ./build.sh
          echo ::set-output name=exit_code::$?
          echo "Exited with $?"
</code></pre>
</blockquote>
</aside>
<p>There are two problems here:</p>
<ol>
<li>
<p>If the <code>build.sh</code> script fails, the step will exit there, so the <code>set-output</code> will never be issued. See <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#exit-codes-and-error-action-preference">Exit codes and error action preference</a> for details and the recommended workaround.</p>
</li>
<li>
<p>The second echo will always write “Exited with 0”, because in Bash <code>$?</code> is always the exit status of the <em>previous</em> command. For the second <code>echo</code> that’s the first <code>echo</code>, and I don’t see how that could fail.</p>
</li>
</ol>
<p>For the comment steps, note that the documentation on <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#job-status-check-functions">Job status check functions</a> says this:</p>
<blockquote>
<p>If your <code>if</code> expression does not contain any of the status functions it will automatically result with <code>success()</code>.</p>
</blockquote>
<p>For the “Create comment if build is ok” step that’s probably fine, but for the “Create comment if build fails” I assume you want a condition like this:</p>
<pre><code class="lang-yaml">if: failure() &amp;&amp; steps.build.outputs.exit_code != 0
</code></pre>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/martyrs">@martyrs</a>,</p>
<p>Glad to see you in Github Community Forum!</p>
<p>You can get the step status via <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#steps-context" rel="nofollow noopener">step context</a>, check the <code>steps.&lt;step id&gt;.outcome</code> value if it’s failure, success…etc.</p>
<p>In the latter step, check with if expression <code>if: steps.build.outcome=='failure'</code> for build fail.</p>
<p>Checked on my side:<br>
<img src="https://user-images.githubusercontent.com/52772761/181095029-e68700a4-76d5-4e4e-939b-c755e4fad5b1.png" alt="image" data-base62-sha1="3FfV47pGz89Lnzrogm1CYp4m84r" width="612" height="212"></p>
<p>Thanks</p>

**Comment by martyrs** [comment]
<p>Thank you all.<br>
I resolved this with the following change:</p>
<pre><code class="lang-auto">      - name: Create comment if build is ok
        if: ${{ success() }}

.........

      - name: Create comment if build fails
        if: ${{ failure() }}
</code></pre>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/martyrs">@martyrs</a>,</p>
<p>Just one comment, <code>${{ failure() }}</code> will return  <code>true</code>  when <strong>any previous step</strong> of a job fails.<br>
Hence, if you use it to check the status for step ‘Build Docker Image’, please make sure only step ‘Build Docker Image’ is possible to fail.</p>
<p>Thanks</p>
